### DISCUSSION TITLE
Controlled Spotlight

### DISCUSSION THREAD

**Discussion by strblr** [mantinedev/mantine#3394] [Feature requests]
Hey,

I really like the spotlight package and would like to use it to make a global search in an app. This requires to fetch the backend and passing the search string. The problem is that the spotlight wasn't thought for this it seems, I haven't found a way to "control" the search string from outside the component.

My current workaround is something like this:

```ts
const [search, setSearch] = useState("");
const items = useFetch(search);

<SpotlightProvider
  actions={items.map(...)}
  filter={(query, actions) => {
    setSearch(query);
    return actions;
  }}
>
```

But it's not optimal. Would be very nice to have additional optional props on `SpotlightProvider`, `value` and `onChange` for example.
Thank you!

**Comment by vood** [comment]
Yes, came here to open a discussion. Would love to be able to pass the query string.

**Comment by rtivital** [comment]
It is planned to introduce controlled spotlight in 6.0

**Comment by malekeym** [comment]
@mantine/spotlight is a very powerful component that we can use in our project, but If the search result is not immediately available and requires an API call, what is the recommended approach? or maybe filtering should be done on the backend side(for example by recommendation system)
 things that I think should change/add:
- add loader indicator beside the search icon
- make the `filter` prop accept a function that returns Promise(the type of `filter` changed to  `SpotlightAction[] |  Promise<SpotlightAction[]>`)
- more control on `Spotlight` by `useSpotlight` hook (for example resetting actions, filtering, or updating.... )

Do we have any update on this? I would be happy to contribute. Please let me know if there's anything I can do to help.

**Comment by viveknigam3003** [comment]
My use case is also similar to @strblr, where I have a step by step spotlight, where onTrigger of one action updates the action list for spotlight, but inside a certain onTrigger I require the `query` value in order to make an API call and fetch the list upon which the user can take action. 

Like for example I have a "Go to" action, which when triggered gives options for "Project", "Template", and "Workspace". Now what I want is to be able to click either of these, and then allow the user to search my DB using the spotlight itself. So if the user is typing something in the query (search box), if I can get the realtime state of it, it will be great, because then I can make an API call with the query to get the list and then create action (list) based on that list.

I saw that the onTrigger function in the action gives the `action` values in v6, but not the query value. So as in the video, I need to make an API call to search workspaces in the third step.

Is there any way I can achieve this? 🤔 

https://github.com/mantinedev/mantine/assets/30192068/610c1be9-6098-43d9-9f44-9833cc047a42

**Comment by limdingwen** [comment]
Hi, I do it like this, hope this example code helps. Sorry for the unfiltered code, but in general the hooks happen in this order: `setQuery -> query -> useDebounce -> debouncedQuery -> useEffect (async API call here) -> results -> Spotlight.actions`.

```
"use client";

import { Spotlight, SpotlightActionGroupData } from "@mantine/spotlight";
import { useEffect, useState } from "react";
import { useDebounce } from "use-debounce";

type ClientFriendlyData = {
  href: string;
  title: string;
};

async function getResults(query: string): Promise<{
  debate: ClientFriendlyData[];
  bill: ClientFriendlyData[];
}> {
  const url = `${process.env.NEXT_PUBLIC_SUPABASE_URL}/functions/v1/search`;
  const headers = new Headers({
    Authorization: `Bearer ${process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY}`,
  });
  const request = new Request(url, {
    headers: headers,
    method: "POST",
    body: JSON.stringify({ query }),
  });
  const response = await fetch(request);
  return response.json();
}

function convertResultsToActions(results: {
  debate: ClientFriendlyData[];
  bill: ClientFriendlyData[];
}): SpotlightActionGroupData[] {
  const debateActions = results.debate.map((debate) => ({
    id: debate.href,
    label: debate.title,
  }));
  const billActions = results.bill.map((bill) => ({
    id: bill.href,
    label: bill.title,
  }));
  return [
    {
      group: "Debates",
      actions: debateActions,
    },
    {
      group: "Bills",
      actions: billActions,
    },
  ];
}

// Empty action groups make nothingFound not work properly, and empty headers ain't that good anyway
function clearEmptyActionGroups(actionGroups: SpotlightActionGroupData[]) {
  return actionGroups.filter((group) => group.actions.length > 0);
}

export default function SearchSpotlight() {
  const [query, setQuery] = useState("");
  const [debouncedQuery] = useDebounce(query, 500);
  const [results, setResults] = useState<{
    searchHappened: boolean;
    actionGroups: SpotlightActionGroupData[];
  }>({
    searchHappened: false,
    actionGroups: [],
  });

  useEffect(() => {
    let isCanceled = false;
    (async () => {
      if (debouncedQuery != "") {
        const results = await getResults(debouncedQuery);
        if (isCanceled) {
          return;
        }
        setResults({
          searchHappened: true,
          actionGroups: clearEmptyActionGroups(
            convertResultsToActions(results),
          ),
        });
      } else {
        setResults({ searchHappened: false, actionGroups: [] });
      }
    })();
    // If the debouncedQuery changes before the request completes, cancel the request's setActions call
    return () => {
      isCanceled = true;
    };
  }, [debouncedQuery]);

  return (
    <Spotlight
      query={query}
      clearQueryOnClose={false}
      onQueryChange={(query) => {
        setQuery(query);
      }}
      actions={results.actionGroups}
      shortcut={null}
      filter={(_, actions) =>
        // Don't filter; search is already done
        actions
      }
      scrollable
      highlightQuery
      nothingFound={
        results.searchHappened ? "No results found" : "Start typing to search"
      }
      radius="lg"
    />
  );
}
```

(Public domain)

**Reply by limdingwen** [reply]
Also, yes, the filtering is done on the backend for me (I even disable client-side filtering as shown above). The main way I interact with `Spotlight` is by changing its `actions` property via the aforementioned React hooks.
